lib/jxl/enc_fast_lossless.cc

```diff
@@ -3657,80 +3748,188 @@ JxlFastLosslessFrameState* LLEnc(const unsigned char* rgba, size_t width,
+                             frame_state->ac_group_data_offset);
   }
 
-  auto run_one = [&](size_t g) {
-    size_t xg = g % num_groups_x;
-    size_t yg = g / num_groups_x;
-    size_t group_id =
-        onegroup ? 0 : (2 + num_dc_groups_x * num_dc_groups_y + g);
-    size_t xs = std::min<size_t>(width - xg * 256, 256);
-    size_t ys = std::min<size_t>(height - yg * 256, 256);
-    size_t x0 = xg * 256;
-    size_t y0 = yg * 256;
-    auto& gd = frame_state->group_data[group_id];
-    if (collided) {
-      WriteACSection(rgba, x0, y0, xs, ys, stride, onegroup, bitdepth, nb_chans,
-                     big_endian, hcode, gd);
+  return frame_state;
+}
 
-    } else {
-      WriteACSectionPalette(rgba, x0, y0, xs, ys, stride, onegroup, hcode,
-                            lookup.data(), nb_chans, gd[0]);
+template <typename BitDepth>
+void LLProcess(JxlFastLosslessFrameState* frame_state, bool is_last,
+               BitDepth bitdepth, void* runner_opaque,
+               FJxlParallelRunner runner,
+               JxlEncoderOutputProcessorWrapper* output_processor) {
+  if (frame_state->process_done) {
+    JxlFastLosslessPrepareHeader(frame_state, /*add_image_header=*/0, is_last);
+    if (output_processor) {
+      JxlFastLosslessOutputFrame(frame_state, output_processor);
     }
-  };
-
-  runner(
-      runner_opaque, &run_one,
-      +[](void* r, size_t i) { (*reinterpret_cast<decltype(&run_one)>(r))(i); },
-      num_groups_x * num_groups_y);
+    return;
+  }
+  constexpr size_t kMaxLocalGroups = 16;
+  bool onegroup = frame_state->group_sizes.size() == 1;
+  bool streaming = !onegroup && output_processor;
+  size_t total_groups = frame_state->num_groups_x * frame_state->num_groups_y;
+  size_t max_groups = streaming ? kMaxLocalGroups : total_groups;
+  size_t start_pos = 0;
+  if (streaming) {
+    start_pos = output_processor->CurrentPosition();
+    output_processor->Seek(start_pos + frame_state->ac_group_data_offset);
+  }
+  for (size_t offset = 0; offset < total_groups; offset += max_groups) {
+    size_t num_groups = std::min(max_groups, total_groups - offset);
+    JxlFastLosslessFrameState local_frame_state;
+    if (streaming) {
+      local_frame_state.group_data =
+          std::vector<std::array<BitWriter, 4>>(num_groups);
+    }
+    auto run_one = [&](size_t i) {
+      size_t g = offset + i;
+      size_t xg = g % frame_state->num_groups_x;
+      size_t yg = g / frame_state->num_groups_x;
+      size_t num_dc_groups =
+          frame_state->num_dc_groups_x * frame_state->num_dc_groups_y;
+      size_t group_id = onegroup ? 0 : (2 + num_dc_groups + g);
+      size_t xs = std::min<size_t>(frame_state->width - xg * 256, 256);
+      size_t ys = std::min<size_t>(frame_state->height - yg * 256, 256);
+      size_t x0 = xg * 256;
+      size_t y0 = yg * 256;
+      size_t stride;
+      JxlChunkedFrameInputSource input = frame_state->input;
+      const void* buffer = input.get_color_channel_data_at(input.opaque, x0, y0,
```

Comment:
At some point, we probably want to add a RAII wrapper around this.

Reply:
Added comment.